mcf-browser/src/back-end/parse.ts

```diff
@@ -0,0 +1,184 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Series} from './data';
+import {ParseMcf} from './parse-mcf';
+import {ParseTmcf} from './parse-tmcf';
+import {ParsingError, ERROR_MESSAGES} from './utils';
+
+ type ParseFileResponse = {
+   /** A list of errors that occurred while parsing
+    * the files
+    */
+   errMsgs: ParsingError[];
+
+   /** A list of the ids of the subject nodes */
+   localNodes: string[];
+
+   /** A list of the generated mcf string for
+    * each CSV row
+    */
+   datapoints: Object[];
+ };
+
+/** Parse files and get the local nodes
+  * @param {Array<Blob>} fileList the list of blobs to be parsed
+  * @return {ParseFileResponse} an object containing all of the ids of the
+  *    subject nodes and the error messages
+  */
+async function getNodes(fileList: Blob[]) : Promise<ParseFileResponse> {
+  const finalReturn: ParseFileResponse = {
+    errMsgs: [],
+    localNodes: [],
+    datapoints: [],
+  };
+
+  // Find TMCF file, if it exists
+  let tmcfFile = null;
+  for (const file of fileList) {
+    const fileName = (file as File).name;
+    const fileExt = fileName.split('.').pop();
+
+    if (fileExt === 'tmcf') {
+      if (tmcfFile) {
+        // If another TMCF file was found, throw an error
+        finalReturn['errMsgs'] = finalReturn['errMsgs'].concat([{
+          'file': (tmcfFile as File).name,
+          'errs': [
+            ['-1', '', ERROR_MESSAGES.MULTIPLE_TMCF],
+          ],
+        }]);
+      }
+
+      tmcfFile = file;
+    }
+  }
+
+
+  for (const file of fileList) {
+    const fileName = (file as File).name;
+    const fileExt = fileName.split('.').pop();
+
+    if (fileExt === 'mcf') {
+      const mcfOut = await ParseMcf.readFile(file);
+
+      if (mcfOut['errMsgs'].length !== 0) {
+        finalReturn['errMsgs'] = finalReturn['errMsgs'].concat([
+          {
+            file: fileName,
+            errs: mcfOut['errMsgs'],
+          },
+        ]);
+      }
+
+      finalReturn['localNodes'] = finalReturn['localNodes'].concat(
+          mcfOut['localNodes'],
+      );
+    } else if (fileExt === 'csv') {
+      if (tmcfFile) {
+        const tmcfFileName = (tmcfFile as File).name;
+        const tmcfOut = await ParseTmcf.generateMcf(tmcfFile, file).then(
+            (mcf) => {
+              const mcfParser = new ParseMcf(
+                  tmcfFileName + '&' + fileName,
+              );
+              return mcfParser.parseMcfStr(mcf as string);
+            },
+        );
+
+        const datapoints = await ParseTmcf.generateDataPoints(tmcfFile, file);
```

Comment:
Looks like this reads the CSV file all over again and parses it.  I wonder if it is possible to piggyback on ParseMcf which reads the file, parses it and creates nodes (https://github.com/datacommonsorg/tools/blob/837c0e7a0cfcda4ae9190480bbac3d05c03b5026/mcf-browser/src/back-end/parse-mcf.ts#L375-L376).

Reply:
Yep, that'll be a change that I make for the performance improvement PR, is it okay to leave it out of this PR for now?

Reply:
ok